In [69]:
import json
import pandas as pd

with open('../Datasets/train.json', 'r', encoding='utf-8') as f:
    json_data = json.load(f)

In [70]:
orig_data = pd.DataFrame(json_data)
orig_data = orig_data.drop(['tags', 'plylst_title', 'updt_date'], axis=1)
orig_data.head()

id                                              songs  like_cnt
0   61281  [525514, 129701, 383374, 562083, 297861, 13954...        71
1   10532  [432406, 675945, 497066, 120377, 389529, 24427...         1
2   76951  [83116, 276692, 166267, 186301, 354465, 256598...        17
3  147456  [394031, 195524, 540149, 287984, 440773, 10033...        33
4   27616  [159327, 553610, 5130, 645103, 294435, 100657,...         9

In [71]:
orig_data_sample = orig_data.loc[:1000]
orig_data_sample = orig_data_sample.explode('songs')
orig_data_sample.rename(columns={'songs':'song_id'}, inplace=True)
orig_data_sample.head()

id song_id  like_cnt
0  61281  525514        71
0  61281  129701        71
0  61281  383374        71
0  61281  562083        71
0  61281  297861        71

In [72]:
orig_data_sample.shape

(45196, 3)

In [85]:
orig_data = orig_data.explode('songs')
orig_data.rename(columns={'songs':'song_id'}, inplace=True)
orig_data.head()

id song_id  like_cnt
0  61281  525514        71
0  61281  129701        71
0  61281  383374        71
0  61281  562083        71
0  61281  297861        71

In [86]:
orig_data.shape

(5285871, 3)

In [22]:
ids = dict()

for l in orig_data['id']:
    if l in ids:
        ids[l] += 1
    else:
        ids[l] = 1

print(len(ids))
print()

for i in ids:
    print(ids[i])

115071

19
42
28
38
53
30
39
20
40
15
157
45
21
66
37
47
31
8
18
16
44
14
18
30
20
45
67
106
21
29
26
14
21
27
40
25
19
21
22
17
58
16
200
27
18
14
24
36
58
45
19
20
45
32
15
30
57
35
81
63
20
46
20
65
43
19
16
15
33
16
56
33
19
29
28
49
37
8
5
23
15
33
26
182
34
32
29
23
19
41
200
30
28
11
60
34
19
9
16
64
24
20
58
116
44
80
16
36
23
44
12
19
19
21
38
17
35
83
52
10
45
154
100
24
8
17
18
16
19
25
20
49
69
13
35
59
8
27
48
65
27
20
14
25
94
93
21
42
33
28
69
18
17
39
18
48
15
34
36
28
39
25
73
42
105
49
24
19
18
39
118
19
10
27
29
9
77
200
114
22
11
126
16
60
10
32
200
19
20
129
37
17
14
44
19
58
15
20
33
21
21
54
35
95
42
55
15
40
22
25
30
21
11
14
14
44
19
16
53
14
40
20
73
187
31
10
25
26
10
20
21
36
41
74
96
12
20
76
25
20
24
36
20
14
9
14
200
8
30
39
38
43
22
200
23
80
91
58
10
27
200
18
26
72
30
23
15
63
16
200
150
22
27
20
64
113
11
10
18
19
61
13
20
60
12
41
13
11
20
19
16
11
19
27
29
18
45
18
14
32
17
16
36
8
20
12
23
73
78
8
39
27
27
161
29
62
51
21
71
12
15
36
19
49
70
20
58

50
25
20
12
12
10
10
70
20
13
24
62
36
5
33
120
39
50
29
5
21
39
71
33
15
54
21
25
102
50
200
13
12
19
93
39
48
31
43
19
20
27
29
147
91
12
12
9
44
9
28
19
40
20
10
64
25
53
26
21
31
23
98
10
10
18
15
90
20
200
100
81
28
25
37
20
27
20
20
18
80
36
22
15
73
22
54
200
200
12
9
12
18
20
18
105
29
32
20
20
178
10
10
10
14
79
10
30
47
11
101
72
38
48
15
27
27
200
16
11
176
47
19
16
48
17
20
122
23
20
29
20
22
157
32
16
15
64
200
8
40
17
15
29
10
63
18
12
14
84
41
93
37
102
33
44
11
28
136
3
48
20
14
11
200
59
28
67
26
20
24
21
15
29
20
15
39
102
38
22
24
40
24
124
22
40
29
25
20
38
111
44
29
34
22
10
18
156
54
30
16
26
12
112
20
12
71
35
14
30
29
18
15
20
159
56
37
18
80
26
74
9
15
108
10
25
136
141
54
35
59
50
33
22
34
44
19
58
25
20
79
51
35
10
117
19
28
135
23
21
19
13
23
82
48
26
76
17
19
20
11
13
49
61
15
43
19
20
41
42
83
11
22
17
39
35
23
14
48
48
124
25
36
105
19
20
22
16
19
11
104
20
24
7
77
50
59
30
23
20
34
37
60
12
10
49
23
26
49
30
119
74
39
13
66
43
11
83
25
20
172
30
28
35
19

175
33
19
41
20
50
20
89
10
200
12
28
42
10
40
62
18
58
127
28
78
15
40
20
38
88
48
23
36
8
32
31
16
13
200
88
76
16
19
29
25
10
48
18
19
38
31
33
51
64
13
27
40
20
20
45
39
27
83
20
58
64
62
15
119
31
23
11
50
20
106
47
35
118
23
50
23
133
42
22
30
117
99
38
47
41
40
10
200
38
12
40
200
43
41
32
20
41
15
71
12
28
31
10
23
200
41
15
49
40
37
115
18
26
24
37
21
22
17
31
200
14
17
78
11
23
200
71
20
21
88
28
39
84
80
19
35
26
97
19
49
17
20
18
55
19
26
40
16
49
61
16
20
93
132
11
14
30
32
25
93
171
15
18
29
30
43
32
47
27
81
43
22
19
25
61
20
80
22
36
47
27
18
33
17
30
33
21
200
200
24
21
35
15
16
110
80
17
10
120
35
200
6
30
200
44
89
73
10
102
34
48
92
1
80
10
93
186
78
27
90
200
68
30
28
19
49
53
43
15
22
32
23
18
19
61
14
135
15
131
30
16
40
22
106
65
18
13
192
37
12
13
142
42
16
26
30
25
192
27
63
87
17
27
8
18
200
11
110
17
14
20
51
13
33
30
24
18
18
15
21
11
50
41
39
50
16
13
23
98
10
44
10
36
34
13
12
20
200
20
20
28
32
39
32
20
144
20
41
24
25
11
49
23
53
31
12
32
200
40
20
19
3

16
38
96
17
22
24
43
13
113
53
200
122
21
21
25
163
73
200
43
44
108
26
50
20
40
43
145
63
20
20
16
20
21
20
82
30
25
71
10
132
78
29
14
20
25
50
23
19
20
64
32
47
20
101
94
21
45
14
32
27
16
14
20
19
200
37
55
200
20
47
15
14
41
43
53
65
57
103
11
59
20
125
10
29
27
25
72
24
18
98
200
30
22
29
20
20
20
81
63
200
20
22
70
77
25
20
47
39
18
26
19
14
21
37
16
93
29
20
20
26
10
200
27
88
17
30
51
40
13
15
77
83
28
20
42
44
45
95
22
18
84
7
32
10
19
18
15
30
69
40
77
15
30
8
32
76
57
25
30
26
37
100
13
103
40
20
87
32
29
12
50
15
51
57
26
36
104
28
10
30
45
10
12
33
19
58
19
20
67
38
200
102
138
92
19
200
36
45
35
31
55
5
47
35
34
23
14
37
10
19
21
16
169
50
15
149
73
62
20
43
51
17
54
24
19
112
16
30
13
72
26
200
36
200
21
50
77
13
84
20
28
130
26
14
53
15
35
66
17
72
18
20
20
20
38
79
32
29
25
200
50
24
29
23
30
101
29
34
88
16
187
200
154
46
98
18
31
16
42
16
49
20
16
17
28
26
31
17
49
125
9
46
97
20
143
79
19
10
20
22
57
10
120
116
28
13
17
21
27
14
69
23
20
20
96
16
63
160
29
5
9
15
1

17
21
23
94
72
22
200
47
20
155
18
28
14
140
47
200
164
30
72
34
10
20
101
43
21
25
200
20
37
20
42
86
85
53
46
19
174
8
9
79
34
121
21
16
18
59
10
19
23
200
8
72
41
45
28
30
38
19
12
73
25
23
25
67
200
200
21
121
18
8
160
22
93
150
19
47
54
23
23
73
19
33
82
15
19
21
48
21
52
23
12
12
71
51
19
47
23
48
11
10
77
29
10
49
200
19
22
170
28
11
26
20
19
30
28
20
200
10
45
26
30
20
43
25
12
29
22
36
23
27
35
200
21
30
147
57
13
89
19
133
19
169
25
20
36
177
41
97
30
31
161
58
52
200
50
69
37
31
32
38
14
15
67
18
19
31
30
62
45
16
11
35
40
36
14
16
31
33
14
81
44
39
81
16
28
31
39
19
54
23
76
118
42
24
33
133
44
200
14
36
200
200
10
23
20
200
15
39
28
185
200
115
100
52
14
18
17
23
17
16
151
25
83
25
20
39
35
22
50
20
32
200
22
27
39
45
19
20
23
18
9
44
29
23
45
11
19
69
16
25
22
94
20
18
60
105
44
16
30
92
20
16
37
22
90
98
20
29
71
74
23
45
5
15
22
17
20
18
84
16
117
50
49
19
25
46
14
18
21
26
30
44
41
28
22
29
68
200
68
17
27
173
18
17
23
50
20
124
40
27
29
17
42
20
72
143
20
64
15
167
30

117
32
36
20
48
24
31
36
41
87
32
15
21
9
5
14
7
130
9
27
18
36
47
7
16
20
14
20
16
26
83
26
30
40
29
49
46
25
16
21
25
18
30
30
16
112
95
20
20
54
18
20
20
13
27
25
10
52
20
63
8
22
9
200
62
18
134
58
45
43
20
20
30
200
50
34
17
93
73
31
59
38
31
22
13
31
27
24
200
27
46
4
27
9
20
47
38
27
69
53
105
25
20
37
133
23
15
27
95
81
31
101
25
65
10
49
26
96
53
15
20
32
24
11
15
17
25
21
17
19
20
20
11
24
28
10
20
39
37
17
155
19
20
152
17
49
26
97
33
19
20
11
21
47
13
161
30
13
10
30
36
9
13
25
42
59
31
27
53
25
18
18
20
42
32
20
93
52
98
28
23
27
38
28
10
37
14
20
6
34
47
34
14
53
21
33
19
83
33
15
45
35
14
109
139
10
17
29
22
33
10
22
31
41
20
18
16
23
191
200
43
25
24
22
47
19
41
43
75
31
50
30
18
200
38
30
17
38
20
27
15
44
95
21
20
19
31
27
15
93
200
200
54
22
67
33
20
27
26
86
19
141
71
107
27
20
15
25
17
19
24
7
10
17
80
58
5
29
11
132
28
199
35
47
24
7
7
37
21
93
18
10
114
12
22
68
26
33
8
27
18
48
100
73
80
200
75
7
33
144
7
12
7
9
153
200
81
43
45
15
38
16
21
26
27
28
17
47
28
25


26
19
54
19
32
30
88
25
57
28
50
30
49
5
17
9
20
67
37
16
9
32
200
106
69
9
18
19
75
14
200
5
22
12
79
54
25
30
15
24
19
75
25
19
12
39
55
12
19
19
40
26
29
23
10
15
26
36
64
93
48
15
96
79
5
7
15
19
23
40
111
5
34
20
200
26
20
24
27
18
20
9
74
32
30
200
197
19
40
56
20
80
26
53
46
33
22
64
15
45
12
147
21
114
30
19
109
73
200
16
76
20
29
26
30
17
20
67
18
26
30
19
15
10
44
87
35
25
40
177
200
45
46
42
39
20
19
14
50
26
35
90
76
107
21
23
42
136
9
200
20
25
47
49
15
50
47
74
112
34
200
8
19
31
35
34
15
18
20
11
13
16
16
125
26
24
89
30
19
49
32
29
20
15
19
68
25
14
40
106
17
10
11
17
12
15
27
73
48
37
53
40
27
71
24
27
33
12
20
29
33
11
15
122
22
93
30
118
28
97
24
106
39
19
30
22
200
77
20
30
21
15
162
29
20
52
45
77
10
20
13
51
66
15
28
45
27
200
22
52
24
63
24
24
62
10
200
15
27
176
74
19
17
19
17
52
73
24
29
30
16
30
9
12
70
200
46
96
58
14
20
25
200
117
25
16
10
30
47
14
24
20
15
94
86
71
58
95
23
15
24
42
105
13
36
30
147
78
66
25
56
128
19
30
37
14
15
28
52
69
22
22
25
10
75
20


22
200
10
21
42
29
24
19
95
34
18
97
32
26
20
31
20
16
26
21
46
43
38
41
178
98
19
122
51
27
26
21
72
44
200
23
71
166
17
12
85
78
53
12
42
115
68
5
14
50
19
22
92
81
10
26
47
133
10
32
40
20
29
92
32
51
20
21
16
18
29
9
19
76
159
38
19
10
28
11
19
106
77
64
144
22
18
46
10
37
11
8
32
28
146
50
17
13
15
21
20
33
30
9
32
16
57
34
25
31
25
59
60
32
32
9
20
81
30
20
11
27
61
155
15
23
141
11
39
43
64
50
15
22
17
11
200
20
20
10
20
19
39
19
24
64
19
54
14
21
25
17
19
92
69
25
65
9
53
17
43
5
98
58
28
57
11
71
20
25
23
9
77
30
29
20
20
41
73
37
142
20
17
10
50
30
69
20
39
34
80
23
76
22
16
200
21
18
24
30
20
63
28
43
29
108
85
20
10
19
107
14
24
51
74
19
22
17
20
37
29
200
20
36
29
52
47
23
29
40
25
200
52
21
38
16
19
89
55
32
19
68
80
30
17
40
46
24
64
20
17
15
34
57
23
60
27
72
161
126
31
14
66
16
26
19
56
58
19
26
52
26
60
29
141
13
51
10
30
25
16
59
86
20
32
18
52
64
16
26
18
82
15
13
42
88
55
32
20
30
15
21
59
91
11
16
15
10
83
14
20
23
20
97
66
16
19
72
35
18
19
14
20
17
56
13
179
70


31
14
200
18
10
61
20
23
69
16
100
7
31
89
31
47
48
78
32
19
10
68
39
95
11
22
108
30
46
35
128
50
15
64
19
40
19
98
121
22
200
11
71
19
123
26
29
60
20
30
20
30
31
36
39
43
30
14
18
27
10
11
21
29
50
38
47
72
44
152
116
54
20
100
162
15
18
56
55
20
83
46
10
20
30
49
97
10
25
48
7
78
20
200
25
36
53
22
25
16
200
92
97
73
29
20
23
38
31
46
30
13
200
35
35
155
20
48
80
104
19
62
74
30
37
24
21
109
76
107
12
19
18
21
27
27
24
54
39
28
34
19
200
99
17
33
55
33
77
22
20
83
13
21
13
65
10
17
33
27
19
32
45
27
16
49
23
35
31
67
36
30
20
23
15
17
9
28
56
56
34
51
11
30
10
11
37
39
21
181
11
22
45
30
27
16
67
33
32
70
14
31
16
22
91
50
10
97
23
20
37
22
36
104
36
59
19
26
139
29
200
34
43
11
200
64
9
18
17
25
37
33
85
8
12
46
27
48
30
10
15
191
12
64
31
20
14
19
12
15
19
8
24
30
20
48
76
19
14
26
6
14
23
32
28
133
14
19
62
52
19
27
51
15
86
93
29
19
10
25
33
128
19
88
30
24
42
45
55
40
32
11
18
52
135
21
200
16
50
13
20
37
23
19
38
25
10
91
38
104
20
30
39
38
67
139
43
42
6
33
20
28
23
22
33
14

15
13
20
20
17
15
16
60
48
11
40
57
10
19
200
22
31
27
21
80
35
30
68
19
57
34
29
92
28
11
18
17
200
24
32
109
20
24
41
26
92
23
39
19
21
13
63
143
15
41
15
12
34
124
21
109
51
18
30
20
35
33
29
20
20
45
44
10
39
17
28
46
23
18
50
16
15
74
39
35
18
54
17
20
30
30
17
37
43
16
31
39
51
24
27
37
50
26
17
12
43
51
56
9
19
24
27
71
49
22
103
27
77
34
82
15
30
25
17
51
31
20
19
112
32
67
33
53
11
46
42
8
73
38
14
93
22
73
69
19
68
78
18
91
15
15
12
28
31
83
8
200
52
20
85
38
14
16
38
47
200
12
41
20
89
27
17
26
30
42
32
25
22
38
20
74
101
13
45
200
83
116
119
18
36
41
68
23
22
46
41
30
44
15
14
33
74
32
94
14
39
5
36
19
29
31
82
24
12
30
30
50
38
10
86
166
90
19
38
19
38
11
151
25
19
5
27
20
50
37
19
72
18
24
44
13
20
98
14
32
20
21
19
57
24
21
40
13
40
86
19
53
82
17
32
60
26
77
20
39
20
11
18
71
61
43
18
32
14
18
17
30
8
25
10
42
13
50
13
50
16
55
39
14
29
10
151
16
20
72
28
20
16
16
24
54
200
48
21
58
10
168
49
27
12
21
30
200
20
33
129
20
63
10
35
28
30
30
41
18
28
147
42
71
126
68
30
30

17
49
19
120
50
55
20
30
40
64
43
29
19
20
20
19
13
46
10
200
30
85
19
30
8
119
28
74
33
200
2
24
25
50
24
33
36
28
18
19
56
11
21
68
19
38
11
139
29
16
19
28
28
50
84
25
23
19
19
35
84
39
50
11
42
35
18
200
12
24
21
51
17
37
26
16
23
200
32
49
16
200
167
26
24
19
19
19
59
46
21
43
20
17
31
30
200
42
36
20
67
22
29
17
19
21
6
100
97
15
31
18
72
32
26
21
10
32
48
37
36
28
34
22
34
48
26
20
40
200
11
43
150
200
24
11
42
75
19
20
14
40
18
34
74
20
16
58
20
30
83
78
94
11
39
48
37
110
20
73
37
158
12
82
66
41
172
54
118
11
17
35
86
23
37
94
93
27
9
29
197
39
23
17
35
10
63
12
31
24
20
33
40
86
27
26
77
20
158
21
128
163
6
31
19
19
16
20
42
12
12
35
44
136
13
58
20
12
26
35
11
67
196
28
11
19
19
38
18
12
41
11
45
23
200
93
9
87
26
80
31
43
16
55
20
25
15
78
42
52
18
52
30
198
55
13
31
20
15
19
18
15
11
45
63
12
37
124
200
17
20
49
28
76
22
76
66
26
21
29
75
15
25
11
25
23
49
25
14
95
71
28
15
56
10
55
18
19
15
10
31
16
27
106
54
187
39
15
16
47
200
39
62
90
123
25
25
51
29
22
200
14
24
30
2

164
22
17
69
13
115
18
155
42
34
13
20
23
17
18
25
61
136
39
24
35
14
41
32
113
31
38
15
17
13
16
174
200
53
71
32
20
49
20
39
11
34
23
16
19
40
79
14
45
17
22
103
20
19
18
49
165
68
39
38
15
82
23
21
14
50
17
18
50
20
46
20
52
55
15
175
13
66
10
11
20
20
14
111
149
61
20
110
24
20
163
23
36
198
117
63
15
34
78
19
29
27
19
200
96
29
47
19
32
67
22
50
37
124
19
50
35
18
14
19
152
20
24
51
12
41
17
28
30
10
20
11
24
15
21
22
49
40
150
40
21
35
50
11
36
23
18
32
42
60
20
94
10
51
39
30
19
76
22
63
43
20
62
27
30
28
10
28
189
111
16
126
49
24
28
16
11
20
94
55
72
44
18
30
9
35
36
26
20
200
17
36
29
88
87
10
140
200
64
30
22
11
20
100
72
13
124
96
38
137
9
30
45
14
158
36
25
75
37
25
30
33
30
31
25
17
24
20
41
200
19
50
49
45
14
22
24
18
105
100
21
38
177
99
15
7
200
54
17
71
141
39
40
13
107
14
16
48
15
22
15
54
200
17
200
49
24
10
13
20
51
25
97
21
47
40
200
112
24
41
200
110
34
40
31
75
57
13
20
74
13
20
83
30
29
20
39
9
28
57
33
107
30
24
21
19
39
50
22
55
21
26
200
25
91
19
9
49
11
200

60
50
30
71
30
54
19
25
32
30
67
34
38
17
155
20
69
39
10
11
200
18
10
24
85
119
18
8
20
90
21
140
167
93
25
102
24
200
14
42
116
26
200
200
91
73
15
21
23
32
24
16
8
19
19
96
12
78
30
38
21
21
69
15
35
2
65
31
55
30
52
20
27
5
23
84
100
29
161
9
200
70
35
30
200
23
200
10
36
44
38
23
37
10
105
171
23
25
21
20
76
33
16
31
32
200
17
9
5
58
18
200
53
20
74
31
60
55
30
200
49
30
15
10
61
41
46
182
10
14
18
200
200
54
69
31
26
31
20
57
10
54
19
48
49
47
10
20
76
29
69
31
200
33
23
17
30
106
21
42
24
35
20
24
86
25
66
39
17
40
34
22
41
20
15
24
52
18
48
35
10
20
5
30
26
34
16
200
75
20
96
20
12
33
41
23
49
51
11
17
29
15
32
26
78
48
10
50
44
10
30
26
36
44
23
12
34
18
20
50
13
56
10
10
47
62
24
13
65
50
31
20
36
33
34
15
67
28
29
10
5
196
16
21
29
21
18
13
20
8
30
45
19
31
51
57
19
30
25
25
21
44
55
28
13
78
36
200
58
7
23
75
28
71
9
43
76
28
26
71
25
55
21
37
17
195
9
40
43
10
19
13
20
20
29
31
49
65
86
68
23
26
19
45
27
21
144
133
20
14
18
38
57
58
50
16
18
47
10
20
21
147
18
10
10
40
23


15
200
106
22
24
31
27
22
26
10
35
20
185
18
31
32
25
12
17
31
91
37
67
37
43
20
73
21
75
47
39
20
40
38
20
17
35
15
54
65
13
17
200
10
6
79
97
16
23
13
21
30
200
110
12
56
35
19
8
156
19
55
20
200
18
76
26
39
166
20
21
91
18
17
9
40
17
50
12
21
11
23
32
27
97
200
30
32
25
20
21
50
151
7
38
27
24
10
13
118
197
19
25
12
26
23
12
20
14
56
5
31
18
29
57
200
27
84
18
41
105
31
61
18
140
37
33
30
26
22
23
19
16
26
49
29
17
23
82
13
76
20
20
19
33
11
19
19
48
44
20
32
18
10
46
23
16
30
22
24
19
47
67
20
20
200
20
40
67
50
17
16
73
30
25
61
22
20
18
38
7
48
34
18
11
27
21
17
16
16
10
35
22
113
34
19
20
31
18
26
19
34
62
70
11
29
25
32
116
44
78
54
18
18
9
41
15
28
15
23
32
73
100
15
200
14
57
20
11
19
13
18
15
25
16
33
30
19
200
20
130
44
32
79
16
94
21
30
24
10
45
53
15
29
75
24
103
21
35
20
24
18
18
17
32
144
16
8
30
30
22
83
8
76
11
50
45
99
86
187
51
29
21
86
55
16
16
191
50
61
20
126
70
17
69
77
17
19
161
37
81
10
10
25
99
13
20
65
17
61
91
28
72
96
111
41
18
46
26
62
61
53
20
43
20
5
19

43
35
24
24
27
111
29
60
95
20
93
131
20
200
38
79
46
44
14
118
41
22
42
41
13
20
20
31
74
11
13
103
29
136
15
7
55
29
21
100
134
70
17
49
32
23
38
53
55
136
20
18
54
12
10
51
20
19
34
25
27
72
15
15
65
27
33
7
18
45
22
48
16
20
47
20
12
11
42
98
21
32
20
20
25
37
30
63
19
18
49
54
19
44
33
77
19
42
184
65
19
29
47
13
46
15
25
46
50
17
29
32
38
30
18
36
42
9
47
49
82
25
80
60
23
42
140
119
12
19
21
24
27
40
30
162
19
15
19
168
59
49
172
30
36
31
21
12
200
73
25
41
73
21
200
29
49
36
57
20
118
33
39
100
29
23
10
124
18
34
15
60
18
200
153
51
134
20
17
19
22
14
20
10
16
20
47
77
14
200
88
23
62
56
15
12
200
14
14
42
20
10
9
47
40
20
200
26
15
21
26
70
62
29
18
14
9
25
11
19
23
82
200
18
16
11
30
54
48
45
18
18
27
15
63
20
10
112
52
58
59
21
51
15
25
20
41
45
16
77
10
30
44
58
200
21
57
51
25
51
30
30
30
24
66
55
18
20
12
19
31
20
36
16
22
17
21
19
25
73
93
44
41
37
119
10
55
29
40
47
42
35
17
20
60
69
17
30
61
23
41
200
30
15
31
38
18
41
41
85
68
14
50
20
46
17
14
27
20
8
21
15
22
125
33

20
38
36
38
16
20
53
16
38
41
45
12
10
21
25
18
21
23
88
39
34
55
17
30
18
32
10
17
102
18
11
20
40
31
37
32
18
87
14
59
28
12
29
85
11
23
62
51
74
29
22
17
23
26
15
48
26
101
39
33
63
33
71
12
10
19
10
41
17
18
30
33
21
32
27
32
103
18
19
20
10
25
30
87
12
18
50
14
200
46
12
26
99
18
15
18
23
12
50
10
17
24
25
48
17
85
56
19
14
200
5
55
20
105
23
14
80
24
26
10
32
28
54
82
23
19
26
34
47
51
16
140
200
13
20
54
153
23
100
8
30
113
19
21
16
36
18
16
127
16
27
10
10
34
53
185
22
43
133
180
23
61
10
29
110
97
30
25
89
23
34
48
92
20
25
19
23
200
56
18
22
24
5
84
37
30
26
26
7
82
21
89
73
135
50
34
14
34
14
200
29
10
20
31
14
56
14
18
20
40
83
18
30
18
74
13
20
24
12
21
18
19
25
47
18
26
27
16
19
55
32
64
24
15
38
41
13
94
85
20
19
29
13
200
73
13
28
13
24
20
20
56
12
20
100
10
20
62
16
19
59
22
31
27
11
40
30
200
161
42
55
13
30
20
31
15
98
22
37
20
20
16
104
32
22
16
10
14
48
50
173
17
29
35
46
104
31
100
10
149
20
34
30
22
49
16
104
29
50
9
200
20
25
20
114
30
41
98
31
20
20
31
53
41
29

18
15
33
18
33
22
27
15
10
23
57
22
103
40
30
12
12
12
41
18
12
200
98
20
18
19
24
47
96
20
22
39
46
15
200
19
10
42
37
59
21
36
18
36
49
24
37
33
125
146
15
39
105
49
200
43
152
119
12
91
10
186
27
30
20
47
8
32
31
73
52
79
15
27
38
22
18
23
20
28
22
200
32
200
13
66
17
10
36
17
34
30
14
22
34
29
54
31
16
120
73
22
142
19
24
45
32
10
56
28
39
20
91
22
58
30
41
91
100
42
19
10
10
42
34
34
30
10
18
26
13
25
19
33
38
25
28
180
45
10
25
16
29
34
9
33
16
19
19
44
50
19
19
38
49
20
14
165
45
15
47
16
95
27
15
21
65
62
48
10
22
18
148
16
37
14
19
18
69
51
49
37
43
29
171
194
177
28
23
66
57
16
19
18
34
33
22
39
59
172
30
24
8
20
76
20
30
82
21
20
45
7
20
39
32
50
27
43
30
54
169
26
16
34
15
200
17
96
71
39
15
34
20
26
16
24
28
32
78
44
200
128
21
27
27
98
26
100
27
24
83
80
31
17
22
48
20
35
19
24
22
125
100
26
20
14
132
42
45
40
134
20
40
12
24
71
84
15
145
23
16
14
22
8
100
107
7
200
19
20
27
55
29
29
41
10
26
79
14
50
20
43
20
20
98
32
42
29
132
50
50
114
37
21
12
200
17
200
40
44
58
18
4

20
22
100
52
19
43
12
20
12
32
5
200
41
48
10
27
34
11
10
109
98
30
45
23
200
25
12
31
18
66
23
29
51
20
25
35
15
40
37
14
17
116
58
50
37
31
30
41
72
29
43
33
166
36
19
60
64
20
22
50
40
25
46
15
11
50
147
24
24
5
23
31
42
25
22
49
35
28
26
10
101
200
25
10
9
23
29
12
91
15
25
200
9
25
27
22
42
24
17
27
200
18
48
22
19
68
26
20
13
86
33
18
20
50
28
120
45
69
17
15
37
48
92
14
30
15
32
90
36
48
20
35
184
67
33
100
29
13
24
40
38
32
27
61
42
24
74
35
39
48
41
34
34
16
54
78
13
64
24
16
14
42
26
21
63
42
41
40
19
52
11
52
37
48
45
96
30
16
68
12
37
20
22
24
98
8
18
20
30
37
14
162
66
33
145
68
19
67
24
19
66
126
13
26
29
26
15
53
56
46
13
29
11
50
88
13
21
79
18
144
57
5
41
124
30
69
34
20
122
200
31
51
19
18
128
16
24
41
37
9
48
17
10
15
49
21
28
20
49
23
25
28
37
45
40
20
127
7
28
30
90
15
87
13
3
26
19
18
39
45
52
20
34
19
21
19
18
98
19
15
23
200
41
9
20
19
24
44
83
26
39
11
108
19
65
21
21
103
13
39
28
20
32
19
100
200
38
89
25
53
10
23
11
161
36
14
42
23
55
17
42
30
23
14
67
76
19


141
42
27
77
11
25
19
29
19
77
35
17
75
17
200
10
46
67
12
16
52
31
32
23
34
23
109
103
30
39
36
20
16
88
26
20
22
64
17
63
74
27
19
23
24
32
46
16
23
97
25
44
10
35
20
21
19
36
55
30
30
3
30
19
34
41
37
27
76
38
12
10
49
21
26
37
28
22
143
14
19
91
33
7
13
13
51
19
17
8
70
78
108
85
25
19
85
46
114
21
200
53
19
85
61
54
20
13
21
65
18
19
10
21
37
52
95
200
20
10
33
28
40
200
77
58
14
19
12
14
24
25
35
44
39
25
27
74
30
65
25
74
29
29
77
6
45
20
128
76
51
25
29
19
81
47
20
40
43
16
35
11
157
30
53
154
30
4
12
23
200
6
116
24
12
28
41
30
15
146
34
10
20
50
17
40
26
22
17
15
20
42
25
27
8
11
14
20
200
20
8
53
15
18
16
41
38
20
64
177
47
29
155
48
73
23
11
29
28
90
27
68
183
200
59
54
100
45
39
200
18
20
28
25
28
67
76
74
101
36
36
13
20
24
20
28
39
37
28
15
16
56
10
200
60
47
80
24
19
30
83
40
80
25
40
21
31
10
63
14
30
200
27
86
25
17
27
42
33
12
147
22
113
116
22
15
28
11
48
30
14
16
20
52
26
41
29
25
89
46
32
68
66
24
73
20
10
109
22
39
200
43
15
61
15
82
38
17
24
5
168
50
170
200
101

37
19
12
66
15
25
200
59
6
22
27
46
19
200
10
90
27
103
10
13
10
17
26
57
35
10
28
50
35
32
16
17
19
41
12
93
23
44
54
46
18
20
17
8
37
29
42
32
23
22
33
10
36
49
31
200
53
99
20
21
49
66
18
16
40
79
25
27
15
74
34
24
91
31
32
25
28
25
20
27
80
20
30
27
17
26
53
30
60
200
27
16
61
22
20
27
108
12
32
20
19
187
14
38
17
61
98
16
27
30
18
15
93
200
29
20
10
110
112
17
22
11
43
24
85
36
28
25
20
13
23
20
59
121
13
10
21
15
15
21
31
84
10
26
25
23
23
24
35
34
27
37
195
37
17
51
54
22
30
13
31
200
68
28
15
10
54
20
30
11
118
49
20
38
27
49
26
34
20
13
98
145
177
72
134
31
68
76
37
16
138
79
23
117
23
44
30
95
41
200
20
111
47
20
66
65
29
65
21
10
30
24
40
20
38
56
17
114
55
41
9
21
96
39
19
36
57
92
13
19
43
200
25
99
83
14
198
93
15
50
104
17
49
41
39
51
120
22
89
30
38
85
18
98
25
95
36
54
22
10
31
20
14
24
18
16
29
21
82
78
45
18
17
24
72
23
128
200
21
24
44
51
23
28
18
27
41
30
24
7
34
75
21
60
30
35
26
18
20
60
20
19
22
66
15
59
93
51
40
31
40
20
8
8
200
29
200
10
55
154
63
27
34
13
50


73
95
24
12
25
19
13
20
23
26
20
139
18
200
12
10
30
16
39
44
26
30
15
65
20
128
74
22
23
200
16
18
29
16
34
56
125
200
28
97
84
13
18
18
31
94
58
189
48
33
28
18
20
10
47
105
10
25
22
95
17
19
9
28
20
36
35
30
32
123
15
12
25
41
26
13
55
71
9
30
30
20
10
46
31
71
18
27
65
44
55
19
5
19
15
55
119
30
200
53
41
200
32
34
18
30
10
20
30
37
41
49
27
17
21
32
15
50
26
20
20
35
8
22
13
163
12
200
13
20
15
170
32
25
56
144
44
31
25
19
15
29
25
24
28
29
32
27
45
45
100
29
58
18
12
51
47
23
26
75
28
17
54
13
42
52
20
87
200
30
32
36
18
34
200
56
30
95
31
14
23
42
49
18
40
47
19
95
149
11
93
38
105
40
18
36
50
29
14
25
21
21
12
33
28
200
24
200
66
12
11
22
19
22
56
18
29
87
15
16
9
52
21
26
15
38
26
12
75
36
51
1
29
25
63
26
11
37
79
19
10
107
10
20
27
104
32
22
77
13
28
47
15
78
109
34
23
16
16
75
43
33
21
32
50
63
53
51
20
24
18
13
19
25
200
8
36
19
11
17
200
9
18
32
37
12
24
25
37
19
20
12
10
60
10
86
30
30
73
20
34
48
43
21
11
13
41
34
33
35
22
200
20
21
73
16
200
42
22
21
23
30
36
41
32
22


48
19
10
31
58
28
61
27
23
30
80
18
19
32
200
54
32
18
54
66
9
22
34
22
97
99
24
160
20
33
30
10
18
30
20
50
10
10
9
30
47
200
113
20
16
10
23
22
23
56
95
200
20
45
19
30
16
24
51
29
5
15
20
20
20
67
58
20
36
19
31
11
12
15
19
142
14
75
200
20
30
19
200
38
46
31
30
30
61
28
56
62
37
18
10
35
16
104
18
41
100
21
14
112
17
19
19
50
20
14
27
16
10
20
36
20
15
35
37
13
19
35
16
156
65
15
12
17
16
60
22
32
88
36
30
19
19
19
47
20
56
29
49
20
35
27
30
35
32
20
44
200
1
20
89
78
45
29
14
28
77
17
44
75
21
15
33
94
36
17
14
50
20
14
15
20
35
8
21
39
112
27
22
22
200
20
104
28
6
62
92
62
15
20
25
34
200
12
29
200
22
39
28
20
30
31
21
15
39
132
125
54
18
40
30
6
200
9
17
10
19
10
36
10
114
24
20
28
27
90
13
12
34
26
6
50
37
17
83
5
17
25
77
21
200
35
197
18
70
28
80
19
183
17
95
39
37
38
38
8
73
83
88
29
79
20
10
33
22
19
20
133
20
31
10
16
45
140
38
38
55
28
37
10
19
20
25
42
36
200
16
37
35
30
11
20
17
39
11
17
87
24
200
17
55
97
46
17
40
18
55
22
16
58
73
149
19
24
18
12
32
58
49
200
200
32
1

88
19
31
5
21
11
10
18
24
11
15
64
75
14
49
29
42
119
33
19
17
40
105
27
46
196
56
72
19
88
31
25
28
16
63
34
27
41
50
27
34
26
45
11
14
29
47
81
21
25
41
12
118
16
20
27
28
74
22
24
15
22
51
27
30
97
31
34
60
46
100
13
10
64
31
20
30
57
20
9
20
18
10
15
12
52
55
19
61
100
200
6
28
78
25
15
22
59
20
20
25
53
13
17
28
30
21
37
30
92
25
13
24
13
29
50
30
25
20
20
42
10
24
42
37
8
106
134
12
10
186
23
29
35
16
31
200
27
54
84
199
20
100
33
187
19
50
26
17
52
48
183
18
25
41
18
16
81
32
29
13
14
72
45
24
29
9
119
31
21
81
22
65
31
33
72
9
64
30
23
26
18
90
19
36
39
50
37
102
32
23
105
30
19
31
34
14
12
17
41
200
24
17
90
20
30
8
48
100
19
16
20
24
9
20
100
19
45
97
20
19
134
31
20
25
31
25
17
5
36
20
19
76
13
13
22
7
10
16
26
79
52
46
71
92
40
114
50
200
17
26
15
16
25
14
67
12
102
6
30
16
20
37
25
17
20
68
74
40
22
40
10
19
10
10
28
22
16
29
18
40
39
44
20
75
11
30
39
107
40
200
20
72
42
42
20
25
18
29
20
61
81
138
98
74
35
25
135
159
15
18
16
17
14
76
15
20
7
16
48
28
20
20
122
35
6
55
2

12
158
27
200
23
105
34
20
49
19
30
73
19
13
28
86
148
65
50
49
20
120
11
52
55
23
40
63
49
75
23
9
29
14
13
42
28
20
33
200
99
18
16
200
37
50
200
25
23
38
46
15
82
20
74
35
31
11
200
51
29
20
13
200
29
22
10
24
26
19
200
6
119
17
26
15
45
18
50
13
80
41
10
92
31
10
12
21
22
53
42
57
31
20
29
28
129
12
63
24
48
20
28
28
21
17
38
7
130
11
106
11
147
22
20
29
19
75
62
52
118
59
52
15
16
136
93
48
35
34
15
34
50
18
28
17
99
38
30
39
14
42
30
17
64
16
126
58
42
25
76
21
12
8
20
13
47
11
18
32
20
23
24
17
21
19
58
141
43
118
82
34
63
31
60
200
11
50
100
100
24
20
40
96
9
10
45
11
15
41
31
40
22
20
58
98
97
29
8
20
100
10
30
19
12
35
19
42
18
52
13
33
100
100
7
73
26
20
9
55
34
10
22
37
20
72
45
18
33
17
28
53
200
40
136
21
191
6
26
61
34
15
15
59
47
20
19
163
15
172
18
200
34
16
20
32
22
165
21
29
60
81
85
25
30
14
33
69
177
25
45
42
12
10
49
16
65
25
28
49
36
36
29
49
40
59
15
200
14
200
26
47
100
24
14
27
20
26
103
18
19
26
26
59
102
56
28
17
30
34
22
31
20
10
36
9
15
16
26
36
31
24
200


64
17
61
15
22
23
54
15
79
27
70
5
18
39
19
31
20
12
19
20
20
15
20
13
67
19
48
10
52
25
122
180
72
73
28
34
93
27
20
39
23
15
16
20
18
20
20
20
21
46
43
52
28
44
89
34
80
72
40
30
29
77
79
19
145
26
98
77
200
18
200
13
21
18
17
106
10
38
9
17
22
35
105
77
18
94
101
21
27
16
78
32
16
63
103
18
19
13
98
10
68
23
41
10
7
36
26
33
200
42
73
20
20
16
19
197
50
28
42
27
10
31
28
35
45
30
20
53
29
22
49
38
23
42
70
22
16
16
48
96
23
26
18
40
20
94
20
74
53
39
89
24
56
25
17
86
33
100
20
20
29
37
62
31
13
38
40
132
23
31
37
21
29
98
12
20
19
50
7
31
24
29
200
24
30
31
98
191
70
27
43
28
27
38
17
18
40
49
30
200
22
25
20
22
22
20
66
25
11
25
20
39
21
26
93
14
56
105
40
33
16
200
38
32
18
58
9
37
35
18
33
38
135
20
124
15
41
29
20
24
33
68
11
23
47
23
6
79
47
61
10
24
15
76
21
56
44
200
86
14
26
18
40
32
22
78
12
29
25
11
30
69
65
10
20
25
106
30
118
24
39
26
200
20
15
83
26
100
18
68
25
200
41
26
28
37
19
22
76
50
31
9
15
51
30
34
43
62
11
18
15
18
166
108
90
30
59
40
13
32
40
29
30
22
44
19
3

16
23
124
25
48
200
21
24
51
15
19
20
5
45
174
74
98
29
24
51
20
13
17
15
43
62
40
97
35
35
11
141
35
94
23
26
5
93
16
200
20
29
71
18
36
18
200
24
100
21
10
26
51
28
15
18
19
25
29
47
11
22
29
200
18
12
72
28
29
25
21
29
44
30
16
22
55
20
22
17
18
33
51
29
11
124
35
30
51
57
21
23
23
21
98
31
22
200
20
63
15
26
75
32
93
87
25
17
34
85
48
26
18
47
42
17
8
25
39
19
26
188
52
19
22
25
15
106
62
83
35
67
44
106
30
15
200
16
37
46
32
6
110
138
13
42
16
38
63
162
27
49
17
14
34
82
30
21
24
10
131
20
31
14
57
13
48
21
22
25
15
30
48
47
27
200
48
138
75
19
62
30
84
28
19
35
24
20
5
13
11
4
12
18
47
35
35
40
65
44
21
22
18
13
100
13
200
24
30
20
18
9
58
19
88
10
15
80
90
77
59
31
20
26
99
28
13
11
46
29
39
88
71
17
124
103
52
58
25
123
38
10
27
19
12
28
108
58
6
128
96
48
75
22
36
59
17
200
41
15
75
200
20
25
20
21
28
28
29
15
27
147
93
22
34
27
19
13
200
9
76
200
22
34
20
176
190
49
143
11
14
25
31
24
35
25
42
182
72
18
34
64
68
48
20
43
22
200
73
20
200
70
148
27
118
25
23
40
50
80
17
14
48


48
20
29
31
50
36
27
12
107
200
16
20
110
60
35
26
160
20
100
23
19
15
23
14
9
98
28
16
22
9
26
66
15
37
50
49
39
45
94
10
30
123
20
68
47
58
18
10
29
24
83
29
28
13
11
52
17
10
89
120
21
71
63
30
25
18
30
32
29
54
158
164
32
14
59
18
104
41
46
78
25
20
86
35
57
31
10
120
36
58
20
41
20
22
36
41
179
9
12
26
37
72
73
200
44
91
41
48
45
20
26
43
13
13
28
20
17
17
18
22
20
19
21
45
30
15
10
30
20
31
102
26
87
6
30
18
30
73
62
103
13
17
18
75
16
57
26
16
19
71
23
36
56
12
28
28
17
75
39
72
96
23
86
52
92
40
30
28
76
12
46
30
200
39
38
42
64
48
21
119
77
30
20
51
48
14
20
55
42
10
24
17
14
200
28
59
19
20
51
16
17
14
64
14
22
26
29
18
10
35
49
59
10
44
106
70
200
95
78
10
17
112
32
63
103
8
36
19
20
20
28
9
12
15
34
35
31
94
21
38
21
20
19
21
16
29
200
36
24
95
24
26
19
52
13
33
11
100
21
42
19
32
10
24
22
14
78
22
48
31
30
35
20
31
24
36
18
40
200
5
46
99
46
200
52
29
17
60
29
60
53
72
15
25
31
13
20
17
33
14
84
69
13
10
23
13
145
61
19
24
200
33
24
15
200
25
30
33
16
63
16
18
200
44
200
6

31
21
37
19
70
26
34
20
89
79
20
24
54
96
61
25
200
19
16
77
15
22
73
39
38
30
33
27
39
200
29
76
11
103
94
16
35
30
36
61
25
22
74
30
84
25
8
32
30
78
26
149
35
21
43
35
64
49
40
10
20
34
102
59
20
41
39
25
17
31
46
28
64
29
100
35
26
8
34
37
20
179
129
27
20
31
35
11
12
32
102
28
30
20
16
77
13
15
26
15
37
30
31
28
33
200
20
50
30
162
139
30
30
116
200
53
104
60
37
25
33
17
25
80
26
29
57
72
22
31
47
73
55
12
21
40
36
24
25
9
45
51
31
10
25
15
85
55
21
17
29
100
38
19
17
45
8
166
98
50
17
20
33
72
53
24
53
16
38
29
100
12
34
90
19
182
19
50
36
45
101
61
86
102
25
16
40
31
55
49
24
77
21
30
79
63
19
21
24
44
35
20
16
8
8
27
63
18
20
126
46
30
13
34
10
28
19
104
31
13
30
72
18
52
41
13
10
19
19
48
81
25
51
15
11
88
11
110
25
75
72
30
20
27
48
16
47
69
10
48
37
44
26
57
28
78
16
44
76
19
62
101
20
36
58
26
15
29
30
10
15
62
89
23
66
19
77
55
99
30
26
22
32
27
104
23
19
38
17
42
24
19
20
20
13
17
29
22
82
20
50
21
86
200
23
30
13
83
50
34
76
7
19
8
16
85
49
15
19
20
52
16
76
29
120
26
56

27
5
31
200
25
19
18
43
23
13
16
42
34
27
200
18
25
68
192
89
55
18
200
25
21
13
19
83
50
30
116
40
30
48
100
37
200
37
17
12
30
200
10
25
62
29
27
20
123
102
16
19
27
65
20
25
111
50
16
57
14
30
55
47
106
54
20
17
39
23
15
101
91
107
27
59
80
10
32
54
84
63
10
34
42
31
15
26
52
33
20
11
62
33
100
83
16
26
14
21
91
64
39
10
36
8
18
57
173
69
45
200
20
37
20
200
48
27
24
50
30
25
10
20
14
24
23
20
16
74
22
17
24
60
17
9
18
170
26
67
47
11
200
27
150
103
104
20
47
36
148
56
33
84
66
30
64
80
34
19
38
18
11
27
34
18
13
200
17
103
19
17
21
125
200
22
53
18
15
9
25
28
53
16
25
33
28
73
87
41
28
37
20
19
81
14
16
20
7
29
14
35
15
18
60
95
17
13
68
11
200
34
45
147
18
83
11
11
58
30
30
20
10
178
25
18
26
14
29
34
25
20
25
19
10
64
26
19
38
29
25
19
20
25
47
94
87
117
11
47
24
18
24
13
36
63
12
12
34
88
20
102
20
18
71
146
60
169
15
51
25
200
33
81
20
20
9
16
6
12
12
94
33
13
43
23
34
22
36
8
200
48
20
47
11
75
200
24
20
15
30
18
17
45
29
17
30
16
60
27
30
21
10
24
20
73
22
31
26
20
11
22
16
5

200
49
17
66
14
57
8
59
18
37
15
39
23
15
26
35
38
20
21
35
10
25
14
27
19
13
35
19
32
38
19
39
25
41
68
42
32
20
18
200
22
10
90
12
18
146
11
25
15
21
9
37
20
54
18
59
13
25
10
91
11
16
21
22
72
26
20
10
21
8
51
10
22
11
100
26
8
89
25
42
60
20
160
73
20
37
76
23
39
2
15
37
72
32
200
50
20
19
28
21
12
200
36
25
22
21
100
48
31
42
14
77
10
23
35
30
18
31
32
21
12
200
22
51
18
19
41
132
21
132
39
16
38
23
142
131
53
20
106
31
46
20
37
62
16
19
32
20
57
47
8
95
49
8
16
85
75
38
70
26
16
48
37
82
20
53
19
9
37
20
116
52
80
11
37
52
15
52
23
8
10
36
30
14
81
50
44
18
41
13
26
131
10
10
20
101
66
78
13
187
41
29
35
9
112
53
33
17
40
16
34
22
54
22
200
35
31
29
134
71
15
21
140
15
19
34
122
15
40
9
31
16
28
12
74
47
27
8
15
41
17
47
33
44
39
22
30
20
54
18
22
89
72
47
87
40
30
14
145
81
28
24
29
38
20
45
36
27
9
33
49
27
18
92
19
50
44
37
18
22
100
19
33
19
54
20
10
19
20
57
32
20
56
131
20
17
29
28
200
27
18
22
25
52
29
19
166
21
25
165
29
16
55
30
10
59
99
14
72
24
21
14
18
30
46
77
20
17


140
24
49
16
27
21
9
9
2
19
18
50
20
116
25
44
21
35
20
10
74
31
14
22
43
20
17
20
12
14
18
39
92
32
84
29
126
31
10
21
31
90
82
111
47
138
12
26
20
40
17
111
25
11
20
30
124
17
52
157
12
39
121
31
28
200
19
19
13
67
28
24
96
30
16
34
37
18
20
10
6
67
10
5
54
22
9
145
135
21
21
20
27
56
19
84
94
15
12
20
16
19
90
32
19
20
20
58
12
24
20
21
12
50
15
18
15
30
32
28
15
29
20
26
24
26
43
44
102
18
12
42
11
22
109
51
142
13
10
50
23
110
84
173
144
12
12
25
50
32
23
63
16
18
87
91
18
13
21
111
23
26
49
66
15
14
25
27
22
44
48
27
19
26
90
20
81
47
56
25
22
20
36
110
42
30
53
74
106
43
24
24
25
27
172
104
18
31
45
200
42
28
29
13
19
20
26
27
73
64
45
18
86
39
20
48
34
21
47
30
65
36
25
26
44
39
66
26
12
37
21
20
200
12
22
33
33
37
28
70
46
200
31
17
110
73
102
45
29
19
45
25
20
25
68
24
28
35
50
53
33
12
200
28
33
49
20
20
25
44
31
20
187
84
52
10
23
200
50
34
16
4
17
14
85
96
105
13
54
11
40
104
26
16
28
18
52
77
25
18
58
90
55
22
26
27
49
16
23
28
39
11
31
28
27
30
51
31
20
23
15
25
18
21
20

41
44
32
16
20
19
18
13
20
48
12
97
22
27
21
65
9
13
25
20
50
12
21
92
200
19
26
59
20
23
155
18
38
37
15
24
19
13
89
45
55
51
29
20
19
10
140
26
101
18
23
33
60
17
29
21
34
45
30
54
30
37
24
200
59
96
23
97
18
5
50
200
113
166
34
18
153
37
93
12
30
20
8
21
29
11
43
36
36
12
74
56
36
9
99
46
46
12
200
60
19
19
28
22
20
11
19
75
93
49
13
19
21
23
84
14
23
19
96
70
47
41
25
80
19
19
44
27
19
25
20
30
71
17
30
200
44
71
46
12
25
29
15
21
9
16
156
20
173
15
33
96
38
40
24
26
39
22
76
41
21
16
21
25
51
103
78
38
30
30
29
65
34
112
20
28
11
17
148
24
81
13
13
19
32
20
15
196
35
33
30
100
145
33
68
51
40
34
45
59
16
46
35
21
27
35
26
21
65
37
69
21
4
18
14
58
38
73
20
31
35
22
31
200
52
128
183
34
60
78
29
14
28
85
81
180
99
111
63
15
21
39
25
20
14
49
200
50
56
30
10
62
51
18
1
55
34
74
30
20
60
75
49
29
18
27
111
38
49
5
50
77
54
59
87
31
20
47
133
153
31
87
40
25
43
12
83
62
7
13
30
85
28
58
50
10
138
30
17
30
19
33
72
129
30
34
70
156
27
16
27
25
97
94
36
15
79
18
7
17
23
7
62
9
15
28
23


25
19
34
25
23
62
128
20
30
55
85
13
145
126
17
33
61
46
20
20
22
27
17
25
15
24
20
33
97
15
66
31
13
47
25
64
18
80
19
47
62
16
10
18
20
48
200
17
19
19
29
37
24
35
17
47
37
14
9
17
19
31
14
133
36
39
29
128
32
37
179
15
200
28
23
31
14
31
14
15
65
19
72
46
17
16
20
58
38
200
27
75
21
33
78
30
89
58
19
58
15
27
30
22
9
43
34
15
20
96
19
105
69
16
29
200
107
107
50
200
22
32
32
14
11
19
83
8
13
55
81
36
70
20
103
37
94
31
15
12
19
20
31
180
22
23
38
16
200
21
52
200
160
10
10
32
30
13
93
200
147
25
16
19
40
20
83
22
32
29
14
21
26
20
9
20
72
15
21
16
34
31
49
119
10
14
182
55
14
20
9
27
99
20
71
35
11
81
45
21
27
18
20
200
32
20
33
23
36
34
34
101
56
11
48
25
18
9
20
13
74
56
13
30
20
118
57
127
25
200
74
18
18
18
37
19
39
18
22
20
166
49
86
26
60
38
20
23
88
31
33
33
43
30
50
27
88
17
20
189
6
15
12
55
40
20
27
30
18
48
15
17
36
25
27
40
41
33
67
30
31
27
15
151
200
200
134
22
17
37
27
32
20
20
33
100
15
19
10
59
60
60
11
16
18
200
52
40
15
40
49
96
12
11
94
24
30
90
14
12
34
64
23
41

40
200
7
15
166
31
21
78
10
22
29
28
27
42
9
84
19
20
62
50
50
98
16
34
68
42
23
200
12
14
156
19
23
110
25
12
28
154
20
25
20
25
18
43
10
95
24
18
28
73
23
24
22
200
20
75
18
26
37
14
27
42
47
21
69
25
24
20
49
18
41
32
43
25
51
51
11
62
42
34
58
54
128
18
16
26
45
20
40
20
75
20
49
111
128
40
27
20
84
30
42
36
34
15
20
60
100
70
113
156
32
7
13
20
20
20
179
48
52
16
26
20
20
20
14
40
61
29
29
47
51
42
18
19
30
18
27
20
13
20
48
42
127
200
16
23
48
58
200
77
30
20
44
90
17
8
34
10
13
21
20
20
19
28
20
83
23
16
48
36
20
21
22
200
37
14
29
169
50
13
51
62
96
41
23
22
26
29
11
20
41
63
20
44
12
37
49
14
21
76
100
124
41
11
51
119
20
18
200
19
12
59
85
64
18
16
48
18
14
50
38
16
29
15
31
37
200
10
89
50
47
84
35
45
29
17
89
62
33
43
32
15
44
36
36
42
20
52
57
59
32
7
13
10
13
47
76
34
19
17
96
20
22
20
47
20
13
15
20
40
24
18
10
36
200
23
19
200
17
28
15
23
29
16
73
80
59
93
42
24
30
44
95
44
36
28
40
16
21
42
18
10
31
153
121
19
16
29
12
25
62
101
37
200
34
11
19
39
20
35
63
45
21
79
18


20
21
24
69
89
18
29
30
21
71
30
22
183
25
106
16
26
57
89
200
18
12
12
20
12
23
26
200
22
13
24
16
25
69
20
17
200
20
55
19
54
199
19
88
27
25
100
200
49
80
46
18
20
89
10
40
31
23
14
12
9
15
24
44
137
28
12
93
6
10
27
101
24
45
28
200
15
118
12
15
18
101
18
30
18
85
38
15
12
45
27
56
100
43
81
169
10
34
39
24
27
39
44
20
19
52
58
85
42
24
200
24
15
29
30
32
30
200
111
11
60
20
13
21
5
20
19
28
23
43
65
14
31
10
24
14
53
14
18
20
19
10
12
50
12
139
14
29
15
101
19
13
27
20
73
14
200
41
23
36
58
43
23
83
29
9
22
22
48
11
25
19
50
20
21
36
16
114
20
8
55
23
35
20
29
5
26
34
10
89
200
156
10
13
19
18
10
16
82
29
40
30
200
34
77
20
14
37
19
19
16
15
50
50
31
20
30
58
12
15
20
16
11
20
140
20
153
96
40
34
11
27
19
36
29
20
5
65
98
55
12
30
89
45
18
26
38
25
200
25
30
55
10
20
58
45
50
35
17
73
43
73
30
42
30
25
20
75
20
14
18
21
30
29
23
40
18
25
19
30
94
25
17
200
24
13
30
13
24
94
20
16
29
19
16
74
76
23
18
40
28
97
31
40
27
12
16
83
13
41
38
57
11
98
37
200
9
15
20
51
20
16
53
36
22
26


55
137
30
24
30
67
200
18
81
15
70
20
23
27
8
200
19
36
8
26
23
19
14
30
50
28
85
99
60
38
20
20
15
16
20
33
60
15
21
74
102
14
200
132
37
30
78
131
17
100
30
41
25
71
21
49
97
99
20
16
91
9
73
19
13
19
22
14
15
200
20
19
53
43
167
10
18
38
28
67
83
20
53
44
29
14
37
39
17
20
39
56
33
39
55
26
15
144
18
93
51
106
200
19
23
24
10
156
54
39
16
21
18
10
11
63
48
30
15
27
20
10
149
21
81
32
92
40
42
18
62
26
42
102
22
19
25
15
35
38
38
43
40
54
200
18
200
13
26
139
88
43
19
20
121
37
35
134
19
63
20
20
63
30
38
24
40
23
21
35
24
52
19
7
5
36
63
11
18
200
105
14
74
79
61
44
50
43
88
39
59
50
200
17
88
57
57
19
18
30
145
26
94
24
19
42
88
6
32
22
25
36
90
10
199
37
19
56
25
46
31
68
33
20
200
20
12
19
32
12
14
30
13
28
104
21
24
29
42
68
29
56
18
10
130
30
73
65
26
19
33
49
19
20
49
19
61
70
20
17
53
72
7
11
24
12
6
15
20
54
22
38
30
53
8
25
12
200
10
11
6
16
29
37
13
20
13
42
23
20
24
60
24
16
73
27
53
29
21
10
30
79
20
200
52
20
12
46
20
200
24
56
10
36
30
35
13
80
59
35
28
29
20
80
49
14


63
20
30
53
29
99
21
30
23
22
53
14
17
30
19
56
33
52
113
73
20
50
42
19
26
44
15
109
15
24
63
34
25
13
200
24
36
24
33
18
97
16
75
53
10
63
67
29
50
22
22
107
23
16
75
18
14
33
43
28
144
34
80
19
29
64
20
17
112
18
31
15
13
165
18
12
59
20
30
18
85
15
7
31
21
22
20
27
118
50
29
25
37
22
38
12
33
35
73
32
37
160
10
96
179
20
54
90
74
22
61
21
21
27
25
13
50
12
57
14
20
16
74
17
18
25
200
173
93
15
18
28
25
19
21
20
15
35
2
13
22
65
10
98
17
18
73
34
200
69
46
20
37
29
130
10
21
35
46
14
24
43
16
55
10
20
21
45
34
30
12
25
28
64
68
145
32
44
13
40
55
92
4
30
83
69
41
41
145
7
17
24
51
24
27
21
66
58
70
26
79
200
50
17
37
29
7
46
105
19
31
37
41
120
19
18
20
64
24
17
12
11
15
48
31
21
24
21
25
17
19
20
16
59
15
14
77
16
23
23
44
13
21
10
9
7
35
200
33
44
9
150
40
28
14
48
18
29
21
18
19
30
26
99
18
39
71
75
143
15
24
18
84
18
200
100
76
55
19
98
26
151
42
15
48
57
47
14
14
53
14
57
23
120
43
50
75
144
90
14
44
55
33
47
15
13
60
20
69
5
19
200
26
54
20
17
31
33
21
27
12
27
14
40
20
12
20


30
27
19
19
33
51
39
105
12
56
71
23
151
31
34
177
51
79
16
34
51
142
11
49
50
37
48
19
107
20
20
9
16
8
96
18
22
38
107
74
117
20
30
38
47
29
12
31
50
15
81
26
62
25
64
30
26
51
73
22
91
20
10
26
21
22
27
52
85
99
18
71
144
39
14
34
63
5
16
20
123
26
76
69
29
19
50
3
10
109
41
56
14
36
22
47
18
11
9
134
38
19
22
62
70
17
26
29
39
12
46
20
22
30
13
19
26
22
28
41
98
12
80
26
26
34
42
22
50
200
8
77
17
90
29
30
32
78
54
28
84
28
15
29
106
23
45
16
16
21
12
46
39
27
15
30
21
19
175
30
32
25
74
100
61
126
25
92
47
19
44
36
20
92
200
54
31
45
20
41
31
56
200
30
96
50
20
56
62
33
14
71
200
23
53
15
13
45
52
34
53
97
20
43
17
200
103
16
16
17
26
23
66
90
10
67
32
20
25
57
19
22
74
37
10
10
88
50
43
14
30
200
13
26
31
26
20
12
29
30
26
72
29
25
83
25
103
17
39
25
9
70
106
111
15
67
65
75
172
51
69
25
27
12
17
17
34
32
44
23
18
8
29
121
28
27
10
11
111
46
25
15
29
38
11
34
58
31
20
29
61
41
27
36
51
11
21
22
15
52
29
20
79
109
58
87
26
56
164
16
20
23
24
16
137
9
22
13
39
167
67
41
5
46
200
11

19
43
19
79
20
76
95
24
19
61
10
27
15
26
17
165
50
19
32
17
9
9
32
46
23
200
54
100
27
200
32
21
10
52
41
26
14
15
10
200
12
68
33
57
10
26
91
37
14
26
38
39
40
38
20
19
200
46
15
19
100
45
84
60
54
18
59
33
56
62
32
36
34
24
11
14
7
31
123
17
106
18
27
15
58
200
38
20
33
20
40
58
53
36
50
71
31
69
24
8
66
26
27
24
13
31
49
11
15
25
18
78
10
22
25
19
108
44
25
20
71
67
17
29
35
52
34
19
20
20
200
30
144
30
30
28
182
75
19
47
200
46
56
14
22
49
200
20
11
41
43
17
10
18
85
19
72
20
96
19
97
35
21
27
37
30
97
17
18
66
76
21
74
7
23
72
18
18
39
37
83
49
35
14
31
20
49
74
71
25
58
28
41
19
20
77
18
40
11
200
28
74
18
11
10
200
24
13
15
13
29
52
41
48
20
36
28
23
4
24
83
19
41
28
93
200
32
9
23
11
36
200
57
5
29
52
45
92
19
75
48
76
200
10
62
25
22
40
13
21
19
37
14
26
30
18
96
44
48
10
22
19
14
39
38
19
16
10
24
10
11
18
26
30
54
20
96
76
24
16
11
34
115
17
73
50
50
197
5
31
13
102
40
29
21
5
40
12
26
48
35
22
43
14
26
19
10
26
91
33
29
20
28
12
96
200
25
30
56
50
33
25
41
20
35
33
113
200

In [37]:
with open('../Datasets/song_meta.json', 'r', encoding='utf-8') as f:
    json_data = json.load(f)

In [38]:
song_data = pd.DataFrame(json_data)
song_data = song_data.drop(['issue_date', 'song_name', 'artist_id_basket', 'album_name', 'artist_name_basket', 'song_gn_gnr_basket'], axis=1)
song_data.rename(columns={'id':'song_id', 'song_gn_dtl_gnr_basket':'genre'}, inplace=True)
song_data.head()

genre  album_id  song_id
0          [GN0901]   2255639        0
1  [GN1601, GN1606]    376431        1
2          [GN0901]   4698747        2
3  [GN1102, GN1101]   2644882        3
4  [GN1802, GN1801]   2008470        4

In [39]:
genre_list = []

for l in song_data['genre']:
    temp = dict()
    g = 0
    mx = 0
    
    for i in l:
        i = i[2:]
        n = int(i) // 100
        if n in temp:
            temp[n] += 1
        else:
            temp[n] = 1
    
    for i in temp:
        if mx < temp[i]:
            g = i
            mx = temp[i]
        elif mx == temp[i] and g > i:
            g = i
        
    genre_list.append(g)

song_data['genre'] = genre_list
song_data.head()

genre  album_id  song_id
0      9   2255639        0
1     16    376431        1
2      9   4698747        2
3     11   2644882        3
4     18   2008470        4

In [56]:
with open('../Datasets/test.json', 'r', encoding='utf-8') as f:
    json_data = json.load(f)

In [57]:
test_data = pd.DataFrame(json_data)
test_data = test_data.drop(['tags', 'plylst_title', 'updt_date'], axis=1)
test_data.head()

id                                              songs  like_cnt
0   70107  [398985, 449403, 411543, 528044, 143048, 98020...         6
1    7461  [196298, 269984, 267805, 175867, 529244, 63825...         0
2   90348  [273433, 331003, 68432, 411659, 117793, 616860...        21
3   58617  [702227, 48152, 440008, 358488, 701041, 540721...         0
4  102395  [630683, 481582, 528550, 285114, 506667, 17922...        38

In [58]:
test_data.rename(columns={'songs':'song_id'}, inplace=True)
test_data.head()

id                                            song_id  like_cnt
0   70107  [398985, 449403, 411543, 528044, 143048, 98020...         6
1    7461  [196298, 269984, 267805, 175867, 529244, 63825...         0
2   90348  [273433, 331003, 68432, 411659, 117793, 616860...        21
3   58617  [702227, 48152, 440008, 358488, 701041, 540721...         0
4  102395  [630683, 481582, 528550, 285114, 506667, 17922...        38

# BorderlineSMOTE

- 너무 많은 데이터로 인해 1000개의 플레이리스트를 뽑아와 oversampling
- 1000개의 플레이리스트를 oversampling한 결과(45196 -> 113467)
- 상당히 과적합으로 보임

In [45]:
from collections import Counter
from imblearn.over_sampling import BorderlineSMOTE

sm_x = orig_data_sample.drop(['id'], axis=1)
sm_y = orig_data_sample['id']

print(Counter(sm_y))

sm = BorderlineSMOTE(random_state=0, k_neighbors=4)
sm_x_res, sm_y_res = sm.fit_resample(sm_x, sm_y)

print(Counter(sm_y_res))
print(sm_x_res.shape)

Counter({131098: 200, 8455: 200, 80395: 200, 31654: 200, 114074: 200, 80036: 200, 27272: 200, 27187: 200, 9180: 200, 141805: 200, 131918: 200, 112794: 200, 42641: 200, 106802: 200, 37260: 200, 110348: 200, 141844: 200, 109650: 200, 127865: 200, 100254: 200, 78149: 200, 67939: 200, 98953: 200, 67951: 200, 50553: 200, 49329: 200, 64447: 200, 70033: 200, 10987: 200, 12221: 200, 68338: 200, 121021: 200, 27449: 200, 2406: 200, 39348: 200, 147065: 197, 145880: 190, 89940: 189, 55459: 187, 95004: 182, 9323: 182, 21520: 179, 3380: 178, 100434: 172, 6762: 171, 32224: 168, 20234: 164, 18218: 163, 6393: 161, 109474: 160, 149053: 159, 31804: 157, 134754: 154, 37064: 150, 5069: 147, 8965: 145, 12066: 144, 50790: 142, 145624: 140, 16182: 135, 131403: 134, 92125: 130, 66532: 129, 45306: 129, 112304: 127, 13041: 126, 50705: 125, 60068: 125, 123055: 123, 120983: 120, 152705: 120, 72590: 118, 19592: 118, 116581: 116, 52701: 116, 112218: 115, 146980: 114, 115660: 114, 15011: 113, 18865: 109, 93888: 107, 

Counter({10532: 200, 76951: 200, 36557: 200, 70741: 200, 10288: 200, 31804: 200, 112060: 200, 89809: 200, 8316: 200, 1516: 200, 5348: 200, 88143: 200, 62696: 200, 136677: 200, 144318: 200, 93836: 200, 43665: 200, 35178: 200, 131098: 200, 61024: 200, 82383: 200, 63703: 200, 137737: 200, 67565: 200, 16037: 200, 137153: 200, 33019: 200, 124665: 200, 118405: 200, 26774: 200, 12956: 200, 141483: 200, 97458: 200, 60192: 200, 95004: 200, 19726: 200, 60834: 200, 98228: 200, 8455: 200, 110924: 200, 93434: 200, 34226: 200, 34198: 200, 140334: 200, 61655: 200, 33025: 200, 28452: 200, 116581: 200, 12743: 200, 132983: 200, 143459: 200, 150721: 200, 18696: 200, 60811: 200, 119639: 200, 134754: 200, 37306: 200, 13366: 200, 15630: 200, 96846: 200, 16542: 200, 1514: 200, 115374: 200, 43222: 200, 50570: 200, 52767: 200, 32533: 200, 140912: 200, 73145: 200, 112195: 200, 10705: 200, 80052: 200, 92503: 200, 130786: 200, 41361: 200, 10901: 200, 133097: 200, 60844: 200, 125058: 200, 51730: 200, 72590: 200, 2

In [46]:
sm_x_res.head(100)

song_id  like_cnt
0   525514.0        71
1   129701.0        71
2   383374.0        71
3   562083.0        71
4   297861.0        71
..       ...       ...
95  556301.0        33
96  655561.0        33
97  534818.0        33
98  695032.0        33
99  516602.0        33

[100 rows x 2 columns]

In [51]:
sm_x_res['song_id'] = sm_x_res['song_id'] // 1
sm_x_res = sm_x_res.astype('int')
sm_x_res.head(100)

song_id  like_cnt
0    525514        71
1    129701        71
2    383374        71
3    562083        71
4    297861        71
..      ...       ...
95   556301        33
96   655561        33
97   534818        33
98   695032        33
99   516602        33

[100 rows x 2 columns]

In [53]:
merge_sm = pd.merge(sm_x_res, song_data)
merge_sm.head()

song_id  like_cnt  genre  album_id
0   525514        71     14   2200223
1   129701        71      9   2201802
2   383374        71     10   2216938
3   562083        71      9     43227
4   297861        71      9    303657

In [55]:
from sklearn.neighbors import KNeighborsClassifier 
kn_sm = KNeighborsClassifier(n_neighbors=1)

kn_sm.fit(merge_sm, sm_y_res)
print(kn_sm.score(merge_sm, sm_y_res))

0.9985987115196489


In [90]:
sm_data = pd.concat([merge_sm, sm_y_res], axis=1)
sm_data.head(100)

song_id  like_cnt  genre  album_id      id
0    525514        71     14   2200223   61281
1    129701        71      9   2201802   61281
2    383374        71     10   2216938   61281
3    562083        71      9     43227   61281
4    297861        71      9    303657   61281
..      ...       ...    ...       ...     ...
95   152422         0      1   2282532  147456
96   152422         0      1   2282532  147456
97   152422         7      1   2282532  147456
98   152422         7      1   2282532  147456
99   152422        10      1   2282532  147456

[100 rows x 5 columns]

In [65]:
import numpy as np

for ti in test_data['id']:
    test_temp = test_data[test_data['id'] == ti]
    
    test_temp = test_temp.explode('song_id')
    
    test_temp_merge = pd.merge(test_temp, song_data)
    test_src = test_temp_merge.drop(['id'], axis=1)
    
    if test_src.empty:
        continue
    
    id_list = kn_sm.predict(test_src)
    
    id_nums = dict()
    
    for i in id_list:
        if i in id_nums:
            id_nums[i] += 1
        else:
            id_nums[i] = 1
    
    ids = sorted(id_nums.items(), key=lambda x: x[1], reverse=True)
    ids_index = 0
    cnt = 0
    
    recommend = []
    
    while cnt < 30 and len(ids) > ids_index:
        for song in sm_data[sm_data['id']==ids[ids_index][0]]['song_id']:
            if not song in test_temp['song_id'] and not song in recommend:
                recommend.append(song)
                cnt += 1
            
            if cnt == 30:
                break
        
        ids_index += 1
    
    print(recommend)

[194542, 148992, 76676, 282168, 131310, 205255, 168353, 466272, 554751, 344769, 678423, 542776, 669027, 671873, 183173, 180280, 537730, 666361, 615236, 581556, 168984, 669453, 544537, 542512, 542667, 25329, 205559, 15327, 46627, 653646]
[681062, 540822, 565196, 565698, 575795, 440779, 405204, 542770, 52626, 65980, 71313, 29624, 439630, 631242, 44668, 269483, 269233, 604560, 661966, 98586, 500751, 177909, 286441, 158289, 637866, 496560, 361240, 553018, 361425, 14316]
[86533, 122836, 124278, 597627, 273433, 686295, 289687, 645990, 136458, 583448, 378627, 324948, 6394, 18821, 41543, 47106, 255469, 86640, 319861, 330740, 188271, 233708, 283895, 527255, 659122, 526963, 368055, 217739, 585053, 598097]
[214891, 376360, 196614, 231791, 106876, 422812, 334672, 14212, 584537, 656832, 56325, 93461, 452637, 499969, 163310, 198982, 545231, 76424, 599676, 274608, 624719, 244008, 188030, 302868, 646751, 142059, 92876, 489138, 368170, 692638]
[122182, 355145, 475313, 471988, 470535, 419853, 596794, 31

[206369, 232046, 421613, 507650, 429671, 399591, 693354, 471515, 537999, 41065, 554292, 120364, 663831, 201587, 319535, 430201, 637799, 257372, 641954, 177126, 402899, 661148, 643821, 347191, 648132, 339988, 342255, 222892, 356379, 354676]
[102206, 361178, 175265, 21690, 476707, 113847, 31258, 547185, 27784, 140749, 421743, 691570, 264929, 483384, 364040, 223411, 338856, 223821, 198358, 282417, 53708, 369900, 83955, 181812, 11908, 205969, 108512, 667098, 579675, 394061]
[251375, 7766, 704752, 580326, 646941, 699557, 157435, 544308, 603044, 347585, 468830, 473385, 476502, 473454, 473275, 473356, 476797, 473432, 473495, 473413, 484233, 467914, 476303, 473490, 473329, 473478, 488894, 473198, 462623, 467848]
[468206, 673851, 433118, 10969, 209932, 524210, 699497, 193269, 471037, 451233, 224130, 436458, 303171, 500484, 223327, 343917, 621707, 25799, 505652, 8659, 582747, 444095, 77783, 657184, 80317, 703820, 645601, 637048, 443907, 443904]
[365133, 543675, 183468, 626851, 294481, 294240, 68

KeyboardInterrupt: 

# ADASYN

- 플레이리스트 내의 노래가 너무 적어 oversampling을 할 수 없음

In [88]:
from imblearn.over_sampling import ADASYN

ad_x = orig_data_sample.drop(['id'], axis=1)
ad_y = orig_data_sample['id']

ad = ADASYN(random_state=0, n_neighbors=1, sampling_strategy='minority')
ad_x_res, ad_y_res = ad.fit_resample(ad_x, ad_y)

print(Counter(ad_y_res))
print(ad_x_res.shape)

ValueError: Expected n_neighbors <= n_samples,  but n_samples = 1, n_neighbors = 2

# OneSidedSelection

- undersampling을 하지만 소수 클래스에 해당하는 플레이리스트의 노래들이 너무 적어 데이터를 너무 삭제함

In [89]:
from imblearn.under_sampling import OneSidedSelection

oss_x = orig_data.drop(['id'], axis=1)
oss_y = orig_data['id']

oss = OneSidedSelection(random_state=42)
oss_x_res, oss_y_res = oss.fit_resample(oss_x, oss_y)

print(Counter(oss_y_res))
print(oss_x_res.shape)

ValueError: Found array with 0 sample(s) (shape=(0, 2)) while a minimum of 1 is required by KNeighborsClassifier.